### ISSUE TITLE
[video_player_web] Integrate Shaka Player for DASH and HLS support.

### ISSUE THREAD

**Issue by ditman** [flutter/flutter#66931]
## Use case

There has been multiple requests to support DASH/HLS videos in the web:

* https://github.com/flutter/flutter/issues/53011
* https://github.com/flutter/flutter/issues/46778

## Proposal

We should explore [Shaka Player](https://github.com/google/shaka-player), which is _an open-source JavaScript library for adaptive media. It plays adaptive media formats (such as DASH and HLS) in a browser, without using plugins or Flash. Instead, Shaka Player uses the open web standards MediaSource Extensions and Encrypted Media Extensions._

## TODO

In order to integrate this with video_player_web, we need a few things:

* Can Shaka's JS be dynamically downloaded when needed? This is to minimize the amount of JS a Flutter web app uses to bootstrap.
* Does Shaka support non-DASH/HLS videos? Can we use it as the core videoplayer? This is to decide how to integrate with the current plugin, or maybe write a new plugin for these types of videos that users optionally use.
* Create a js-interop layer of shaka's [API](https://shaka-player-demo.appspot.com/docs/api/index.html). AFAIK, nobody has integrated Shaka in Dart before.

**Comment by joeyparrish** [comment]
> * Can Shaka's JS be dynamically downloaded when needed? This is to minimize the amount of JS a Flutter web app uses to bootstrap.

Yes, absolutely.  The JS and CSS (if using the UI-enabled build) are available from a various CDNs with versioned URLs, including one that is Google-owned: https://developers.google.com/speed/libraries/#shaka-player

If not using the UI, the JS simply needs to be fully loaded before calling the `shaka.Player` constructor.  See tutorial: https://shaka-player-demo.appspot.com/docs/api/tutorial-basic-usage.html

If using the UI build and UI features, the JS and CSS need to be fully loaded first.

If using the UI with a declarative DOM-based setup, the app needs to wait for the `shaka-ui-loaded` event on document before referencing `video.ui` to get access to the controls and player instances.  See tutorial: https://shaka-player-demo.appspot.com/docs/api/tutorial-ui.html

If using an API-driven setup, the app can call the `shaka.ui.Overlay` constructor as soon as the JS is loaded.  There is no tutorial for this yet.  See API docs (https://shaka-player-demo.appspot.com/docs/api/shaka.ui.Overlay.html) and pending issue to add an API-driven UI tutorial (https://github.com/google/shaka-player/issues/2655).

> * Does Shaka support non-DASH/HLS videos? Can we use it as the core videoplayer? This is to decide how to integrate with the current plugin, or maybe write a new plugin for these types of videos that users optionally use.

Yes.  For convenience, Shaka Player will pass single-file progressive videos through to the platform.  A single API can be used for both adaptive content like DASH & HLS and simple progressive content like MP4s or WebM.  (Note that not every browser supports WebM, but almost all support MP4.)

**Comment by ditman** [comment]
> the app needs to wait for the `shaka-ui-loaded` event on document before referencing video.ui to get access to the controls and player instances.

@joeyparrish is this event thrown when the basic shaka js loads, or only when the UI library loads? Is there a `shaka-loaded` event? Is it enough to wait for the `load` event of the script tag? (we've had [issues](https://github.com/flutter/plugins/commit/dc284c47e39bcc04b2003349fd5d8fe3db7d39e9) with this in the past 😄)

**Comment by joeyparrish** [comment]
No, the Shaka library, without a UI, doesn't do anything on load, so it's ready to use right away.

The UI, however, will scan the document for `data-shaka-player` and `data-shaka-player-container` attributes to create the UI elements and attach JS objects to them automatically.  So if you use our UI, and you use the DOM-based auto-setup feature I just described, the `shaka-ui-loaded` event tells you when that's done.  Then you can access the auto-created player and controls objects from the video element like so:

```js
const ui = videoElement.ui;
const controls = ui.getControls();

const player = controls.getPlayer();
const video = controls.getVideo();
```

If you use the UI, but not the DOM-based setup, you would instead create the UI like this:

```js
// "local" because it is only for local playback.
// When casting, a player proxy object must be used to control the remote playback session.
const localPlayer = new shaka.Player(videoElement);
// "Overlay" because the UI will add DOM elements inside the container,
// to visually overlay the video element
const ui = new shaka.ui.Overlay(localPlayer, videoContainerElement, videoElement);

// As with DOM-based setup, get access to the UI controls and player from the UI.
const controls = ui.getControls();

// These are cast-enabled proxy objects, so that when you are casting,
// your API calls will be routed to the remote playback session.
const player = controls.getPlayer();
const video = controls.getVideo();
```

This snippet is essentially what needs to be added to our UI tutorials in https://github.com/google/shaka-player/issues/2655.

**Comment by ditman** [comment]
@joeyparrish this is *great*, that second snippet is exactly what flutter would need. The problem that I can foresee with the auto-setup is that flutter platform views in the Web are rendered inside a ShadowDOM, and wouldn't be visible to the auto-setup. (We're working on fixing that too, but it is what it is for now 😁)

**Comment by joeyparrish** [comment]
Yes, we always anticipated that the convenience of DOM-based setup wouldn't work for some frameworks.  Glad to hear that these snippets are helpful!  Please let us know what else we can do to help.

**Comment by hbengali** [comment]
Since TypeScript defs had come up in a prior discussion (outside of GitHub) w.r.t Shaka Player integration, I wanted to share that Shaka Player recently shipped a new release with official TypeScript definitions:

https://groups.google.com/g/shaka-player-users/c/w9_fwrFvHGA/m/CreMUAOrAQAJ

**Comment by AriasBros** [comment]
Hi all,

I am writing a package to integrate Shaka Player with Flutter. The initial version can be found here: https://github.com/AriasBros/flutter_video_player_web_shaka

For the people interested in use it as the default video_player_web package:

```
dependency_overrides:
  video_player_web:
    git: https://github.com/AriasBros/flutter_video_player_web_shaka.git
```

For now it is necessary modify the index.html of your application to add the Shaka script by yourself:

```
<script src="https://cdnjs.cloudflare.com/ajax/libs/shaka-player/4.1.0/shaka-player.compiled.js"></script>
```

or, from Google API CDN, an older version for now:

```
<script src="https://ajax.googleapis.com/ajax/libs/shaka-player/4.0.1/shaka-player.compiled.js"></script>
```

@ditman @joeyparrish I am trying to implement lazy load of the Shaka script in this branch: 

https://github.com/AriasBros/flutter_video_player_web_shaka/tree/feat/shaka-lazy-load

But after the `load` event of the script tag, I am getting that `shaka` namespace is not in the scope. Even trying to access to the `shaka` namespace using the browser console gives me an `undefined` as result. In the network tab of the browser I can see how the request to download the script is done successfully.

Could you give me some feedback about this to try a fix?

@ditman Once the lazy load is fixed, I would like to send a PR to the video_player plugin to use Shaka Player as the default web player. It would be my first contribution so I need to pass all the process to send the PR.

---------------------

Outside of the scope of this issue, in this branch I have implemented audio track selection:

https://github.com/AriasBros/flutter_video_player_web_shaka/tree/feat/track-selection

Following the interface defined in this other issue:

https://github.com/flutter/flutter/issues/79079

But the idea is to keep this work outside of the first PR sent with the Shaka integration.

----------------------

Other work I will do in the next days is to play DRM content. For a future hypothetical PR too.

**Comment by ditman** [comment]
Thank you very much @AriasBros!! This is great!

A few comments:

> For now it's necessary to modify the index.html...

I think that is fair, people already have to tweak their index.html for **Google Maps**, [docs](https://pub.dev/packages/google_maps_flutter_web#modify-webindexhtml), because their JS SDK doesn't have any "onload" callback that tell us when the javascript has been *parsed* (and not just "loaded"). (Google Sign-In, does have an "onLoad" mechanism in the script URL. [Docs](https://developers.google.com/identity/sign-in/web/reference).)

> But after the load event of the script tag, I am getting that shaka namespace is not in the scope. Even trying to access to the shaka namespace using the browser console gives me an undefined as result.

I don't know the Shaka loading mechanism very well, but as I said above, the JS needs to be *parsed* by the browser (not just downloaded) for you to be able to access the `Shaka` player. Looking at your code I'd try the following:

1. Inject the script tag it in the `head`, not [the body](https://github.com/AriasBros/flutter_video_player_web_shaka/blob/feat/shaka-lazy-load/lib/src/shaka_video_player.dart#L57).
2. Do not `defer`; I don't think you want to defer this script :) [Code](https://github.com/AriasBros/flutter_video_player_web_shaka/blob/feat/shaka-lazy-load/lib/src/shaka_video_player.dart#L55).
3. Inject the Shaka JS in the plugin initialization code, not when it is RIGHT to play a video (near the [`registerWith` call](https://github.com/AriasBros/flutter_video_player_web_shaka/blob/feat/shaka-lazy-load/lib/video_player_web.dart#L21)). This gives convenience to the user, doesn't block flutter from loading, but also gives it some time to load, before the app starts attempting to create instances.

If the lazy loading doesn't work, it's not the end of the world IMO.

> I would like to send a PR to the video_player plugin to use Shaka Player as the default web player.

Don't worry too much about that; we can promote your video package to those people who need more advanced features than what a raw `<video>` tag can provide (DRM, HLS and tracks, etc...). You can own it and stabilize it for a while, without having to worry about the "firehose" of issues that might come from being the endorsed implementation of the plugin.

Once the plugin is mature, we can take a look at how to continue to maintain it (move it to flutter/plugins, or to somewhere the Shaka team owns, or... you decide to own it forever and we don't need to move it anywhere!)

Again, thanks for the contribution! I hope my message helped!

**Comment by AriasBros** [comment]
Thanks @ditman, very appreciated your answer.

About the lazy load, it is working now. It was a problem with the loading system from Shaka:

```
if(typeof exports!="undefined")
  for(var k in exportTo.shaka)
    exports[k]=exportTo.shaka[k];
else if(typeof define=="function"&&define.amd)
  define(function(){return exportTo.shaka});
else 
  innerGlobal.shaka=exportTo.shaka}
```

As we can see above, Shaka checks if RequireJS is being used, if it is, then it loads the library with RequireJS.

So I did it in a way that Flutter loads the library using RequireJS. I also included a fallback in case that RequireJS is not available anymore, just in case if RequireJS is removed as a dependency in the future. I am  not sure about the necessity of this fallback or not, to be honest. I don't want abuse of your help, but any advice is very welcome.

**Comment by AriasBros** [comment]
This comment will contain some off-topic outside of the scope of this issue, my apologies for that. If some team member mark this comment as hidden, I understand it. Thanks.

> Once the plugin is mature, we can take a look at how to continue to maintain it (move it to flutter/plugins, or to somewhere the Shaka team owns, or... you decide to own it forever and we don't need to move it anywhere!)

The company where I am working is betting a lot with Flutter. We have ported this year our first (web) application (http://fantasy.formula1.com/). And now we are working in an OTT video application, with several requirements that the current video_player plugin (in any flavor, not only web) doesn't meet sadly.

Some of those requirements are:

- Audio track selection.
- Multi angle camera selection.
- Fairplay DRM support.
- Widevine DRM support.
- HLS TS support.
- HLS CMAF support.
- DASH MP4 support.

We would like to implement these requirements for Web, Android and iOS; and we would like to do it in a way that they can be promoted (in the future) to be the official Flutter implementation, not only as a payback to all the work that Flutter framework saves us, but also because in that way it is a way to save an hypothetical migration to the Flutter implementation of these features when they land :sweat_smile:

For these reasons, I would like to know where I should go now, to implement each of these requirements, because a lot of them will require add new methods to the interface of the video_player.

For example, I saw this issue open to track "Support video, audio and text track selections": https://github.com/flutter/flutter/issues/79079

Should I continue in that issue to get feedback from Flutter team and share my progress about that requirement?
or should I open a new issue for each of the requirement I want to implement?

**Comment by kv0871916** [comment]
@AriasBros  how to add licenseUri

**Comment by AriasBros** [comment]
Hi @kv0871916, sorry, but for now there is not DRM support. It is something that I will start to work in the next days.

The only work done right now is the Shaka player integration itself and the ability to change the audio track following the interface done in this other issue: 

https://github.com/flutter/flutter/issues/79079

To use multi language you will need to use other dependency overrides:

```
dependency_overrides:
  video_player:
    git:
      url: https://github.com/AriasBros/flutter_plugins.git
      path: packages/video_player/video_player
      ref: feature_track_selection
  video_player_platform_interface:
    git:
      url: https://github.com/AriasBros/flutter_plugins.git
      path: packages/video_player/video_player_platform_interface
      ref: feature_track_selection_platform_interface
  video_player_web:
    git:
      url: https://github.com/AriasBros/flutter_video_player_web_shaka.git
      ref: feat/track-selection
```

**Comment by AriasBros** [comment]
@kv0871916

Here you have the branches with Widevine DRM support (Other DRMs based in Server URL license should work too)

```
dependency_overrides:
  video_player:
    git:
      url: https://github.com/AriasBros/flutter_plugins.git
      path: packages/video_player/video_player
      ref: feat/drm-support
  video_player_platform_interface:
    git:
      url: https://github.com/AriasBros/flutter_plugins.git
      path: packages/video_player/video_player_platform_interface
      ref: feat/drm-support
  video_player_web:
    git:
      url: https://github.com/AriasBros/flutter_video_player_web_shaka.git
```

**Comment by ditman** [comment]
> So I did it in a way that Flutter loads the library using RequireJS. I also included a fallback in case that RequireJS is not available anymore, just in case if RequireJS is removed as a dependency in the future. I am not sure about the necessity of this fallback or not, to be honest. I don't want abuse of your help, but any advice is very welcome.

@AriasBros have you tried compiling your app in "release" mode? I don't think "requireJS" is available there at all (and you shouldn't trust it being there, or use it directly). See this:

* https://github.com/dart-lang/sdk/issues/33979

> Should I continue in that issue to get feedback from Flutter team and share my progress about that requirement?
or should I open a new issue for each of the requirement I want to implement?

No, issues are free to create, so please open (at least) an issue about "The future of the video_player plugin", where you list all the features that you know that you're going to need, and your plan to execute on those. I see two possible outcomes here (there's probably more): either the existing plugin gets expanded in all the directions needed, or you guys end up creating/owning an `drm_video_player` plugin that ends up being the preferred one by the community, and we end up pointing people to using your video player instead of `video_player` :)

I don't know if you've seen the `Plus plugins` (https://plus.fluttercommunity.dev), but those were plugins originally created by the flutter team, that are now owned and maintained by the community!

**Definitely** create a new issue so this can be discussed there!

<details>

<summary><tt> ...fantasy.formula1.com... </tt></summary>

![](https://i.kym-cdn.com/photos/images/original/000/884/384/ec6.jpg)

(PS: Love it!)

</details>

**Comment by nateshmbhat** [comment]
@ditman @joeyparrish @stuartmorgan-g @AriasBros — I'd like to pick this up and carry it through to a landed implementation. Reviving the thread with what's changed since 2022.

## Why now

Track selection has landed natively while this issue sat idle:

- **Audio track selection** — flutter/packages#9925, flutter/packages#10171, flutter/packages#10312, flutter/packages#10313 (Android + iOS)
- **Video track / quality selection** — flutter/packages#11474, flutter/packages#11475, flutter/packages#11476 (Android + iOS)
- **DRM (Widevine + FairPlay)** — flutter/packages#11115, currently in review

`video_player_web` implements none of it. It's still a bare `<video>` element with direct `src` assignment and no MSE, so all six track methods throw `UnimplementedError` on web.

That gap can't be closed with native browser APIs:

- `HTMLMediaElement.audioTracks` is Safari-only — Chrome, Edge and Firefox all ship it disabled behind a flag
- Video quality selection has **no** native web API at all; Safari's native HLS deliberately doesn't expose variants
- DASH plays natively in no browser
- HLS is Safari + Chromium 142+ desktop, and never desktop Firefox

So web is now the only platform where these features cannot work, and the gap grows with each feature that lands.

## What Shaka unlocks

| Capability | Native today | Web today | With Shaka |
|---|---|---|---|
| HLS | ✅ | Partial | ✅ all browsers |
| DASH | ✅ Android | ❌ | ✅ |
| Audio track selection | ✅ | ❌ | ✅ |
| Video quality selection | ✅ | ❌ | ✅ |
| DRM — Widevine / PlayReady / FairPlay | 🚧 #11115 | ❌ | ✅ |
| Live + low-latency HLS/DASH | ✅ | Partial | ✅ |
| Offline + persistent licenses | ❌ | ❌ | ✅ |
| Thumbnails / trick-play | ❌ | ❌ | ✅ |

Embedded captions (CEA-608/708, in-manifest TTML) also become possible, but that needs new platform-interface surface, so I'd keep it out of scope here.

## Why Shaka rather than hls.js

Shaka passes progressive MP4/WebM through to the platform, so it can be a single code path for every source instead of branching progressive vs. adaptive. It also does DASH, which hls.js structurally cannot, and covers all key systems through one EME config — hls.js supports DRM only for fMP4, not MPEG-2 TS. The cost is payload: ~266 KB gzipped vs ~116 KB for hls.js-light.

## Proposed approach

Building on @AriasBros's prototype and @ditman's guidance from 2022:

1. Lazy-load Shaka from CDN at plugin registration, opt-in — apps that only play MP4 keep today's zero-JS path.
2. Progressive sources keep working unchanged, via Shaka passthrough.
3. Implement the six track methods over `getVariantTracks()` / `selectVariantTrack()` and the audio language APIs.
4. DRM last, matching whatever config shape #11115 settles on.

## Open questions

- Is CDN-loaded third-party JS acceptable in an endorsed plugin — and should it be opt-in or the default?
- Fold into `video_player_web`, or a separate package endorsed later, as @ditman suggested in 2022?
- On Safari: keep native HLS (which forfeits track control) or force MSE? iOS Safari needs 17.1+ for Managed Media Source.

Related Issue : #53011

Happy to write this up as a design doc first if that's the preferred path.

**Comment by stuartmorgan-g** [comment]
> Happy to write this up as a design doc first if that's the preferred path.

That would be the best way to determine whether there is buy-in from the relevant owners. (/cc @tarrinneal FYI)

> * Fold into `video_player_web`, or a separate package endorsed later

This is ultimately up to `team-web` to decide, but FWIW my personal recommendation would be for this to be a community plugin-in, not part of `video_player_web` and not the endorsed implementation. Currently what the Flutter-team-owned `video_player` implementations provide access to is the playback functionality that is part of the host system, not any third-party player, and that's what `video_player_web` currently provides. A wrapper around a third-party player on web is comparable to existing third-party implementations of `video_player` based on things like media kit or FFmpeg.

Expanding the scope of the Flutter-team-owned implementation from "wraps host-provided functionality" to "attempts to polyfill across platforms using third-party libraries" would be a large, and permanent, expansion in plugin scope.

**Comment by ditman** [comment]
> Is CDN-loaded third-party JS acceptable in an endorsed plugin — and should it be opt-in or the default?

This is done for google sign-in and google maps web, off the top of my head. See for example:

* https://github.com/flutter/packages/blob/main/packages/google_identity_services_web/lib/src/js_loader.dart
* https://github.com/flutter/packages/tree/main/packages/google_maps_flutter/google_maps_flutter_web#setup

Your JS should come from a legit-looking location.

(Also, what Stuart said about endorsing. This'd definitely be a good separate plugin, but don't build it with the intention of it being endorsed.)

**Comment by nateshmbhat** [comment]
Design doc is written up, per @stuartmorgan-g's suggestion above:

- Doc: https://flutter.dev/go/video-player-web-shaka
- Design doc issue: https://github.com/flutter/flutter/issues/192464

It follows the standard template and covers packaging, architecture, the `dart:js_interop` surface, script loading, how the six track methods map onto Shaka's APIs, the Safari/iOS matrix, DRM (deferred to a later phase), phasing, testing and migration.

On the two points raised in this thread:

**Packaging.** The doc recommends a community-owned, unendorsed federated package (`video_player_web_shaka`) rather than folding Shaka into `video_player_web`. That matches @stuartmorgan-g's reasoning that the Flutter-team-owned implementations wrap host-system playback, and @ditman's "don't build it with the intention of it being endorsed". The doc lays that argument out rather than pushing back on it, and notes that nothing in the design would foreclose team-web revisiting the question later.

**CDN-loaded JS.** Thanks for the `google_identity_services_web` and `google_maps_flutter_web` pointers — the loader in the doc has the same shape: a pinned, versioned URL per plugin release, a `window.shaka` check first so apps can self-host or bundle the script under a strict CSP, and a fallback to today's bare `<video>` path if the script fails to load, so a CDN outage degrades playback rather than breaking the app.

The open questions I'd most like input on:

1. CDN versus shipping the compiled script (~266 KB gzipped) as a package asset, for offline/PWA apps and strict-CSP environments.
2. Safari policy — native `src=` HLS tolerates streams that MSE cannot fetch for CORS reasons, so adopting the package must not break a stream that plays today.
3. Whether the synchronous `isVideoTrackSupportAvailable()` should stay `false` while the script is still loading, or track-method availability should be gated behind `initialize()`.

Comments welcome directly on the doc. cc @tarrinneal @joeyparrish @AriasBros

**Comment by stuartmorgan-g** [comment]
> Design doc is written up, per @stuartmorgan-g's suggestion above:
> [...]
> The doc recommends a community-owned, unendorsed federated package

To clarify: you only need a Flutter design document if the goal is to convince the Flutter team to own this functionality. If you are fine with a community-owned, unendorsed plugin, then you don't need a design doc at all, and can just build it like any other third-party package. Unendorsed implementations do not require any buy-in from the Flutter team.

So if the doc's recommendation is that this not be built or owned by the Flutter team, it's not clear what feedback you would be looking for from Flutter team members.

**Comment by ditman** [comment]
@stuartmorgan-g IMO this issue should have the "would be a good package" label (not sure how widely that is used though?)

**Comment by stuartmorgan-g** [comment]
> IMO this issue should have the "would be a good package" label (not sure how widely that is used though?)

Not very much any more because of [this issue hygiene guidance](https://github.com/flutter/flutter/blob/master/docs/contributing/issue_hygiene/README.md#closing-issues) about when to close issues:

>> it is a feature request that we are unlikely to ever address, and if we did address it, it would not be part of the core SDK (e.g. it would be in a package). (For example, anything in the [would be a good package P3](https://github.com/flutter/flutter/issues?q=is%3Aopen+is%3Aissue+label%3A%22would+be+a+good+package%22+label%3AP3) list is a good candidate for closing without fixing.)

Basically, if it would be accepted as a feature in `video_player_web` then it should be tracked as a regular issue, and if it wouldn't be then the right thing to do would be to close it, not to keep it open with more labels.

**Comment by nateshmbhat** [comment]
Thanks @stuartmorgan-g, @ditman and @mdebbar for the clear answers, and @joeyparrish for the Shaka notes on the doc. My goal was Flutter-team ownership, and the doc framed that badly. I understand that's out of scope for video_player_web, so I won't pursue it further here. Fine to close this issue per the hygiene guidance.

**Comment by ditman** [comment]
I'll revert the design doc go link, since this is not moving forward!

* https://github.com/flutter/website/pull/13896

For posterity, the doc is still up here:

* https://docs.google.com/document/d/1WfP1Ji7Xe-2zbYH6GFeweX6J8tvOVLryZbEU-4NtyBc/edit?usp=sharing
